### PR TITLE
Cache items by group ans speed up group-related operations

### PR DESCRIPTION
**Pull request by bell07**
This change does cache all group -> items relations at start in lua table and speed up all recipes operations. This gives performance boost for progressive mode.
The function `craftguide.get_group_items(str)` I exposed to get access to this cache from outside the craftguide. I plan to use them in my progressive-revealed mod

### CODE REVIEW

File: init.lua [hunk 1]
```diff
@@ -135,23 +133,54 @@ local function item_has_groups(item_groups, groups)
...
 
-local function groups_to_item(groups)
-	if #groups == 1 then
-		local group = groups[1]
-		if group_stereotypes[group] then
-			return group_stereotypes[group]
-		elseif reg_items["default:" .. group] then
-			return "default:" .. group
+local function get_group_items(str)
+	local group_string = str:sub(7)
```
Comment (kilbith) [inline init.lua]: Use the name `group`

File: init.lua [hunk 2]
```diff
@@ -135,23 +133,54 @@ local function item_has_groups(item_groups, groups)
...
+	if group_items[group_string] then
+		return group_items[group_string] --group_items[group][name] = def
+	end
+	local groups = group_string:split(",")
+	local list
```
Comment (kilbith) [inline init.lua]: Surround these variables with newlines.

File: init.lua [hunk 3]
```diff
@@ -135,23 +133,54 @@ local function item_has_groups(item_groups, groups)
...
+			for item, def in pairs(list) do
+				if not group_items[group][item] then
+					list[item] = nil
+				end
+			end
 		end
 	end
+	if list and not next(list) then
+		list = false
+	end
```
Comment (kilbith) [inline init.lua]: Surround this block with newlines.

File: init.lua [hunk 4]
```diff
@@ -135,23 +133,54 @@ local function item_has_groups(item_groups, groups)
...
+	return list
+end
+craftguide.get_group_items = get_group_items
 
-	for name, def in pairs(reg_items) do
-		if item_has_groups(def.groups, groups) then
-			return name
-		end
+local function groups_to_item(str)
+	local group_string = str:sub(7)
```
Comment (kilbith) [inline init.lua]: Use the name `group`.

File: init.lua [hunk 5]
```diff
@@ -135,23 +133,54 @@ local function item_has_groups(item_groups, groups)
...
+	elseif reg_items["default:" .. group_string] then
+		return "default:" .. group_string
 	end
 
-	return ""
+	local group_items = get_group_items(str)
+	if not group_items then
+		return ""
+	end
+	for item, def in pairs(group_items) do
```
Comment (kilbith) [inline init.lua]: Newline above.

File: init.lua [hunk 6]
```diff
@@ -524,32 +552,27 @@ end
 
 local function progressive_show_recipe(recipe, inv_items)
 	local inv_items_size = #inv_items
-	for _, item in pairs(recipe.items) do
+	for _, item_string in pairs(recipe.items) do
```
Comment (kilbith) [inline init.lua]: Keep `item`.

File: init.lua [hunk 7]
```diff
@@ -616,6 +639,10 @@ local function get_init_items()
 			c = c + 1
 			init_items[c] = name
 		end
+		for group, _ in pairs(def.groups) do
```
Comment (kilbith) [inline init.lua]: Remove `, _`

File: init.lua [hunk 8]
```diff
@@ -524,32 +552,27 @@ end
...
 					item_in_inv = true
 				end
 			end
 		end
-
 		if not item_in_inv then
 			return
 		end
 	end
-
```
Comment (kilbith) [inline init.lua]: No.

File: init.lua [hunk 9]
```diff
@@ -524,32 +552,27 @@ end
...
 		else
 			for i = 1, inv_items_size do
-				if inv_items[i] == item then
+				local inv_item = inv_items[i]
+				if inv_item == item_string or inv_item == reg_aliases[item_string] then
 					item_in_inv = true
 				end
 			end
 		end
-
```
Comment (kilbith) [inline init.lua]: No.

**Comment by pauloue** [comment]
I think this is needlessly complicated. The `group_stereotypes` table is already used as a cache, just add more groupstring-->item mappings to it whenever a new one is resolved in `groups_to_item`. `get_group_items` and `group_items` are not needed (we only need one item, that matches the groupstring, not all of them).

Also, I suggest using Luacheck. It reports several bugs.

**Comment by bell07** [comment]
@kilbith, how did you the benchmark? I thestet with the whynot mods collection (new version published yet) and feels the difference, but do not have any benchmark values.

@pauloue, the group_stereotypes are just for get an representative item for a group to be shown in UI.
This cache is for faster resolving groups like `group:flower,color_red` to get all possible items to faster usages calculation and to faster progressive calculcations.

File: README.md [hunk 10]
```diff
@@ -66,3 +66,16 @@ function craftguide.progressive_filter_recipes(recipes, player)
 	return filtered
 end
 ```
+
+### Cache Helper tools
+
+#### `craftguide.get_group_items(str)`
```
[the remark this answers is not in the document]
Comment (bell07) [reply README.md]: This function is to reuse the new `group_items` cache table. Without the function any other mod needs to build up own cache.

File: init.lua (hunk 1 above)
Comment (bell07) [reply init.lua]: `group` is used bellow for resolved group. I need 2x variable names for `group:flower,color_red` and `flower`. Maybe other variable name?

File: init.lua (hunk 4 above)
Comment (bell07) [reply init.lua]: same as above. This group string is not resolved and can be 'flower,color_red'. `group` should be used for resolved groups only.

File: init.lua (hunk 6 above)
Comment (bell07) [reply init.lua]: same as above. The value is not cleaned. The value is not checked if it is registered item, registered alias, unknown item or group string

**Comment by kilbith** [comment]
If you can't benchmark your code, this is not serious. I don't care about feelings, I care about data.

**Comment by bell07** [comment]
Hm, you are right, my feel was outdated. I was convinced that new hash table should get the performance boost so I did not a benchmark :-/ 

Now I benchmarked  `get_progressive_items()` and `get_item_usages()` Result is sobering, there is no boost, just a small improvement which does not justify a new big cache table :-( Maybe my implementation needs more optimization I do not see currently?

Upstream version:

```
- Full inventory - progressive
bench get_progressive_items	0.307002
bench get_progressive_items	0.247309
bench get_progressive_items	0.28344899999999
bench get_progressive_items	0.34322
bench get_progressive_items	0.34267699999998
bench get_item_usages	0.14338799999996
bench get_item_usages	0.18894
bench get_item_usages	0.19577900000002
bench get_item_usages	0.13819899999999
bench get_item_usages	0.14081799999997
bench get_item_usages	0.17968300000001
bench get_item_usages	0.13141199999995
bench get_item_usages	0.150793
bench get_item_usages	0.137565
bench get_item_usages	0.13419300000001

- Empty inventory - progressive
bench get_progressive_items	0.22190700000002
bench get_progressive_items	0.32030499999999
bench get_progressive_items	0.455297
bench get_progressive_items	0.20180400000001
bench get_progressive_items	0.31795599999998

- Prorgessive disabled
bench get_item_usages	0.21247499999998
bench get_item_usages	0.22117500000002
bench get_item_usages	0.28052400000001
bench get_item_usages	0.20388500000001
bench get_item_usages	0.28032300000001
```

My version

```
- Full inventory - progressive
bench get_progressive_items	0.24002900000005
bench get_progressive_items	0.24308499999995
bench get_progressive_items	0.33904299999995
bench get_progressive_items	0.34184299999993
bench get_progressive_items	0.23163699999998
bench get_progressive_items	0.34269600000005
bench get_item_usages	0.17935799999998
bench get_item_usages	0.22512600000005
bench get_item_usages	0.16827799999999
bench get_item_usages	0.14325300000007
bench get_item_usages	0.15673500000003
bench get_item_usages	0.17923600000006

- Empty inventory - progressive
bench get_progressive_items	0.19492300000002
bench get_progressive_items	0.23205099999996
bench get_progressive_items	0.19488100000001
bench get_progressive_items	0.36649299999999
bench get_progressive_items	0.16545399999995

- Prorgessive disabled
bench get_item_usages	0.127344
bench get_item_usages	0.134959
bench get_item_usages	0.143638
bench get_item_usages	0.135097
bench get_item_usages	0.13208
bench get_item_usages	0.19215000000001
bench get_item_usages	0.128412
bench get_item_usages	0.141422
bench get_item_usages	0.209862

```

**Comment by bell07** [comment]
I admit this is the second valueless PR in order from my site. Since the caching did not adduce the expected performance boost, this PR can be closed without merge.

Sorry for the inconvenience.

**Pull request closed without merging**